pkg/monitors/elasticsearch/query/response.go

```diff
@@ -0,0 +1,174 @@
+package query
+
+import (
+	"encoding/json"
+)
+
+// Represents all kinds of supported aggregations. aggregationResponse type can
+// be determined in the following manner -
+// 1) single bucket aggregationInfo contains only "doc_count" and a map to sub aggregations
+// 2) multi bucket aggregationInfo contains "buckets"and a map to sub aggregations
+// 3) metric aggregationInfo contains "value" or "values"
+// These structs are defined based on inputs from
+// https://github.com/elastic/elasticsearch/tree/master/server/src/main/java/org/elasticsearch/search/aggregations
+type aggregationResponse struct {
+	// Number of documents backing the aggregationInfo
+	DocCount *int64 `json:"doc_count,omitempty"`
+	// Non nil for single value metric aggregations
+	Value interface{} `json:"value,omitempty"`
+	// Non nil for multi-value metrics aggregations
+	Values interface{} `json:"values,omitempty"`
+	// Map of sub aggregations with aggregationInfo names, includes both bucket
+	// and metric aggregations
+	SubAggregations map[string]aggregationResponse `json:"-"`
+	// Non nil for multi-value bucket aggregations
+	Buckets []*bucketResponse `json:"buckets,omitempty"`
+	// This field only exists to handle cases where buckets are returned as a
+	// map from the key to the object itself. In such cases the "buckets" field
+	// will be post processed and put into the "Buckets" field of this struct
+	BucketsMap map[string]*bucketResponse `json:"buckets,omitempty"`
+	// All other key-value pairs
+	OtherValues map[string]interface{} `json:"-"`
+}
+
+// Represents a slice ("bucket") of ES documents
+type bucketResponse struct {
+	// Value uniquely identifying a bucket for a given bucket aggregationInfo
+	Key interface{} `json:"key,omitempty"`
+	// Reported in cases where buckets are grouped by a non string key,
+	// for instance time
+	KeyAsString string `json:"key_as_string,omitempty"`
```

Comment:
Just make this a method called `KeyAsString` that does the conversion instead of storing it on the struct directly.  This doesn't actually come from a real ES response as far as I know.

Reply:
`KeyAsString` is returned by ES in cases where the bucket is identified by a non-string value like a date. Since these keys are dimensions for us, I'm considering looking at the type of `Key` and proceed only if the key is of string type. What do you think?

Reply:
I've updated to look at buckets only when the key is a string